README.rst
playsound
=========
*Pure Python, cross platform, single function module with no dependencies for playing sounds.*

Installation
------------
Install via pip:

.. code-block:: bash

    $ pip install playsound

Done.

If you insist on the (slightly) harder way of installing, from source,
you know how to do it already and don't need my help.

The latest version of the source code can be found at:
https://github.com/TaylorSMarks/playsound

Quick Start
-----------
Once you've installed, you can really quickly verified that it works with just this:

.. code-block:: python

    >>> from playsound import playsound
    >>> playsound('/path/to/a/sound/file/you/want/to/play.mp3') 

Documentation
-------------
The playsound module contains only one thing - the function (also named) playsound.

It requires one argument - the path to the file with the sound you'd like to play.

There's an optional second argument, block, which is set to True by default. Setting it to False makes the function run asynchronously.

On Windows, uses windll.winmm. WAVE and MP3 have been tested and are known to work. Other file formats may work as well.

On OS X, uses AppKit.NSSound. WAVE and MP3 have been tested and are known to work. In general, anything QuickTime can play, playsound should be able to play, for OS X.

On Linux, uses ossaudiodev. I don't have a machine with Linux, so this hasn't been tested at all. Theoretically, it plays WAVE files. Any other file format working would surprise me. The block argument is not used on Linux - I do not know whether the function blocks on Linux or not.

Requirements
------------
I've only tested playsound it with Python 2.7 on Windows 7 and OS X 10.11, but
I expect that it should work on Windows XP thru 10 at least, OS X 10.5 and newer,
all versions of Linux, and Python 2.3 and newer.

Copyright
---------
This software is Copyright (c) 2016 Taylor Marks <taylor@marksfam.com>.

See the bundled LICENSE file for more information.


CHANGES
1.2.0 (Feb. 20, 2016)

    * Now uses   NSSound.initWithContentsOfURL_byReference_ on OS X
      instead of NSSound.initWithContentsOfFile_byReference_. This allows it to
      handle both local file paths and URLs. This capability was already
      unintentionally present on Windows by using windll.winmm. Since I lack a
      Linux machine to test on, and nobody is offering assistance, many things
      that work well on OS X and Windows probably don't work at all on Linux.

1.1.0 (Jan. 26, 2016)

    * Changed from using winsound.PlaySound on Windows to windll.winmm.
      Appears to be just as common as winsound, and has the advantage of playing
      more file formats, such as MP3, than winsound (which only plays WAVE as
      far as I can tell).

1.0.0 (Jan. 24, 2016)

    * Initial commit

playsound.py
class PlaysoundException(Exception):
    pass

def _playsoundWin(sound, block = True):
    '''
    Utilizes windll.winmm. Tested and known to work with MP3 and WAVE on
    Windows 7 with Python 2.7. Probably works with more file formats.
    Probably works on Windows XP thru Windows 10. Probably works with all
    versions of Python.

    Inspired by (but not copied from) Michael Gundlach <gundlach@gmail.com>'s mp3play:
    https://github.com/michaelgundlach/mp3play

    I never would have tried using windll.winmm without seeing his code.
    '''
    from ctypes import c_buffer, windll
    from random import random
    from time   import sleep

    def winCommand(*command):
        buf = c_buffer(255)
        command = ' '.join(command).encode()
        errorCode = int(windll.winmm.mciSendStringA(command, buf, 254, 0))
        if errorCode:
            errorBuffer = c_buffer(255)
            windll.winmm.mciGetErrorStringA(errorCode, errorBuffer, 254)
            exceptionMessage = ('\n    Error ' + str(errorCode) + ' for command:'
                                '\n        ' + command +
                                '\n    ' + errorBuffer.value)
            raise PlaysoundException(exceptionMessage)
        return buf.value

    alias = 'playsound_' + str(random())
    winCommand('open "' + sound + '" alias', alias)
    winCommand('set', alias, 'time format milliseconds')
    durationInMS = winCommand('status', alias, 'length')
    winCommand('play', alias, 'from 0 to', durationInMS)

    if block:
        sleep(float(durationInMS) / 1000.0)

def _playsoundOSX(sound, block = True):
    '''
    Utilizes AppKit.NSSound. Tested and known to work with MP3 and WAVE on
    OS X 10.11 with Python 2.7. Probably works with anything QuickTime supports.
    Probably works on OS X 10.5 and newer. Probably works with all versions of
    Python.

    Inspired by (but not copied from) Aaron's Stack Overflow answer here:
    http://stackoverflow.com/a/34568298/901641

    I never would have tried using AppKit.NSSound without seeing his code.
    '''
    from AppKit     import NSSound
    from Foundation import NSURL
    from time       import sleep

    if '://' not in sound:
        sound = 'file://' + sound
    url   = NSURL.URLWithString_(sound)
    sound = NSSound.alloc().initWithContentsOfURL_byReference_(url, True)
    sound.play()

    if block:
        sleep(sound.duration())

def _playsoundNix(sound, block = True):
    '''
    Utilizes ossaudiodev. Untested. Probably works with all version of Linux
    with Python 2.3 or newer.

    Inspired by, and more or less copied from, Bill Dandreta's post on
    this mailing list (since deleted, so I link to a web archive instead):
    https://web.archive.org/web/20080218155209/http://mail.python.org/pipermail/python-list/2004-October/288905.html
    '''
    import ossaudiodev
    from sys  import byteorder
    from wave import open as waveOpen, AFMT_S16_LE, AFMT_S16_BE
    
    with waveOpen(sound, 'rb') as sound:
        channelCount, sampleWidth, framerate, frameCount, compressionType, compressionName = sound.getparams()
        try:
            from ossaudiodev import AFMT_S16_NE
        except ImportError:
            if 'little' in byteorder.lower():
                AFMT_S16_NE = ossaudiodev.AFMT_S16_LE
            else:
                AFMT_S16_NE = ossaudiodev.AFMT_S16_BE
        data = sound.readframes(frameCount)

    speaker = ossaudiodev.open('/dev/dsp', 'w')
    speaker.setparameters(AFMT_S16_NE, channelCount, framerate)
    speaker.write(data)
    speaker.close()

from platform import system
system = system()

if system == 'Windows':
    playsound = _playsoundWin
elif system == 'Darwin':
    playsound = _playsoundOSX
else:
    playsound = _playsoundNix

del system